File: Android_Nativo_Java/Projeto GS300 Java/app/src/main/java/com/exampledv/exemplogs300_java/fala/LanguageSettings.java
package com.exampledv.exemplogs300_java.fala;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.os.Build;
import android.os.LocaleList;
import android.util.DisplayMetrics;
import android.util.Log;

import androidx.annotation.RequiresApi;

import com.google.gson.Gson;

import java.lang.reflect.Method;
import java.util.Locale;

public class LanguageSettings {

    private static final String TAG = LanguageSettings.class.getName();
    private static final String LOCALE_FILE = "LOCALE_FILE";
    private static final String LOCALE_KEY = "LOCALE_KEY";

    /**
     * Obtém o idioma preferido do usuário salvo nas preferências.
     */
    public static Locale getUserLocale(Context context) {
        SharedPreferences preferences = context.getSharedPreferences(LOCALE_FILE, Context.MODE_PRIVATE);
        String localeJson = preferences.getString(LOCALE_KEY, "");
        return jsonToLocale(localeJson);
    }

    /**
     * Obtém o idioma atual do sistema.
     */
    public static Locale getCurrentLocale(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            return context.getResources().getConfiguration().getLocales().get(0);
        } else {
            return context.getResources().getConfiguration().locale;
        }
    }

    /**
     * Salva o idioma preferido do usuário nas preferências.
     */
    public static void saveUserLocale(Context context, Locale userLocale) {
        SharedPreferences preferences = context.getSharedPreferences(LOCALE_FILE, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = preferences.edit();
        String localeJson = localeToJson(userLocale);
        editor.putString(LOCALE_KEY, localeJson);
        editor.apply();
    }

    /**
     * Atualiza o idioma no nível do aplicativo.
     */
    public static void updateLocale(Context context, Locale newUserLocale) {
        if (needUpdateLocale(context, newUserLocale)) {
            Configuration configuration = context.getResources().getConfiguration();
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                configuration.setLocale(newUserLocale);
            } else {
                configuration.locale = newUserLocale;
            }
            DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();
            context.getResources().updateConfiguration(configuration, displayMetrics);
            saveUserLocale(context, newUserLocale);
        }
    }

    /**
     * Verifica se é necessário atualizar o idioma.
     */
    public static boolean needUpdateLocale(Context context, Locale newUserLocale) {
        return newUserLocale != null && !getCurrentLocale(context).equals(newUserLocale);
    }

    /**
     * Serializa um objeto Locale para JSON.
     */
    private static String localeToJson(Locale userLocale) {
        if (userLocale == null) return "";
        Gson gson = new Gson();
        return gson.toJson(userLocale);
    }

    /**
     * Desserializa um JSON para um objeto Locale.
     */
    private static Locale jsonToLocale(String localeJson) {
        if (localeJson == null || localeJson.isEmpty()) return Locale.getDefault();
        Gson gson = new Gson();
        return gson.fromJson(localeJson, Locale.class);
    }

    /**
     * Altera o idioma do sistema (Requer permissões especiais).
     */
    @RequiresApi(api = Build.VERSION_CODES.N)
    public static void setSystemLanguage(Locale locale) {
        if (locale == null) return;
        try {
            LocaleList localeList = new LocaleList(locale);
            changeSystemLanguage(localeList);
        } catch (Exception e) {
            Log.e(TAG, "Erro ao alterar o idioma do sistema: " + e.getMessage(), e);
        }
    }

    /**
     * Método interno para alterar o idioma do sistema (API interna).
     */
    @RequiresApi(api = Build.VERSION_CODES.N)
    private static void changeSystemLanguage(LocaleList localeList) {
        try {
            Class<?> activityManagerNativeClass = Class.forName("android.app.ActivityManagerNative");
            Method getDefault = activityManagerNativeClass.getDeclaredMethod("getDefault");
            Object iActivityManager = getDefault.invoke(activityManagerNativeClass);

            Class<?> iActivityManagerClass = Class.forName("android.app.IActivityManager");
            Method getConfiguration = iActivityManagerClass.getDeclaredMethod("getConfiguration");
            Configuration config = (Configuration) getConfiguration.invoke(iActivityManager);

            config.setLocales(localeList);
            Method updateConfiguration = iActivityManagerClass.getDeclaredMethod("updatePersistentConfiguration", Configuration.class);
            updateConfiguration.invoke(iActivityManager, config);
        } catch (Exception e) {
            Log.e(TAG, "Erro ao alterar idioma do sistema: " + e.getMessage(), e);
        }
    }
}

File: Android_Nativo_Kotlin/texttospeaks/src/main/java/br/com/estudos/texttospeaks/LanguageSettings.kt
package br.com.estudos.texttospeaks

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Configuration
import android.os.Build
import android.os.LocaleList
import android.util.DisplayMetrics
import android.util.Log
import androidx.annotation.RequiresApi
import com.google.gson.Gson

import java.util.*

object LanguageSettings {

    private const val TAG = "LanguageSettings"
    private const val LOCALE_FILE = "LOCALE_FILE"
    private const val LOCALE_KEY = "LOCALE_KEY"

    fun getUserLocale(context: Context): Locale {
        val spLocale = context.getSharedPreferences(LOCALE_FILE, Context.MODE_PRIVATE)
        val localeJson = spLocale.getString(LOCALE_KEY, "") ?: ""
        return jsonToLocale(localeJson)
    }

    fun getCurrentLocale(context: Context): Locale {
        val locale: Locale
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            locale = context.resources.configuration.locales[0]
        } else {
            locale = context.resources.configuration.locale
        }
        return locale
    }

    fun saveUserLocale(context: Context, userLocale: Locale) {
        val spLocal = context.getSharedPreferences(LOCALE_FILE, Context.MODE_PRIVATE)
        val edit = spLocal.edit()
        val localeJson = localeToJson(userLocale)
        edit.putString(LOCALE_KEY, localeJson)
        edit.apply()
    }

    private fun localeToJson(userLocale: Locale): String {
        val gson = Gson()
        return gson.toJson(userLocale)
    }

    private fun jsonToLocale(localeJson: String): Locale {
        val gson = Gson()
        return gson.fromJson(localeJson, Locale::class.java)
    }

    @Suppress("DEPRECATION")
    @RequiresApi(Build.VERSION_CODES.N)
    fun updateLocale(context: Context, newUserLocale: Locale) {
        if (needUpdateLocale(context, newUserLocale)) {
            val configuration = context.resources.configuration
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                configuration.setLocale(newUserLocale)
            } else {
                configuration.locale = newUserLocale
            }
            val displayMetrics = context.resources.displayMetrics
            context.resources.updateConfiguration(configuration, displayMetrics)
            saveUserLocale(context, newUserLocale)
        }
    }

    fun needUpdateLocale(context: Context, newUserLocale: Locale): Boolean {
        return newUserLocale != null && !getCurrentLocale(context).equals(newUserLocale)
    }

    @RequiresApi(Build.VERSION_CODES.N)
    fun setSystemLanguage(locale: Locale) {
        Log.d(TAG, locale.language)
        changeSystemLanguage(LocaleList(locale))
    }

    @RequiresApi(Build.VERSION_CODES.N)
    fun changeSystemLanguage(locale: LocaleList) {
        if (locale != null) {
            try {
                val classActivityManagerNative = Class.forName("android.app.ActivityManagerNative")
                val getDefault = classActivityManagerNative.getDeclaredMethod("getDefault")
                val objIActivityManager = getDefault.invoke(classActivityManagerNative)
                val classIActivityManager = Class.forName("android.app.IActivityManager")
                val getConfiguration = classIActivityManager.getDeclaredMethod("getConfiguration")
                val config = getConfiguration.invoke(objIActivityManager) as Configuration
                Log.d(TAG, "changeSystemLanguage()  LocaleList: " + Gson().toJson(locale))
                config.setLocales(locale)
                val clzParams = arrayOf<Class<*>>(Configuration::class.java)
                val updateConfiguration = classIActivityManager.getDeclaredMethod("updatePersistentConfiguration", *clzParams)
                updateConfiguration.invoke(objIActivityManager, config)
            } catch (e: Exception) {
                Log.d(TAG, "changeSystemLanguage()  Exception: " + e.localizedMessage)
                e.printStackTrace()
            }
        }
    }
}
